src/Microsoft.IdentityModel.Protocols.WsFederation/Configuration/WsFederationConfigurationValidator.cs

```diff
@@ -0,0 +1,195 @@
+                {
+                    ErrorMessage = LogMessages.IDX22708,
+                    Succeeded = false
+                };
+            }
+
+            if (!Uri.IsWellFormedUriString(configuration.TokenEndpoint, UriKind.Absolute))
+            {
+                return new ConfigurationValidationResult
+                {
+                    ErrorMessage = LogMessages.IDX22709,
+                    Succeeded = false
+                };
+            }
+
+            if (configuration.SigningKeys == null || configuration.SigningKeys.Count == 0)
+            {
+                return new ConfigurationValidationResult
+                {
+                    ErrorMessage = LogMessages.IDX22710,
+                    Succeeded = false
+                };
+            }
+
+            // Get the thumbprint of the cert used to sign the metadata
+            string signingKeyId = string.Empty;
+            var signatureX509Data = configuration.Signature.KeyInfo.X509Data.GetEnumerator();
+
+            if (signatureX509Data.MoveNext())
+            {
+                var signatureCertData = signatureX509Data.Current.Certificates.GetEnumerator();
+                if (signatureCertData.MoveNext())
+                {
+                    X509Certificate2 cert = null;
+
+                    try
+                    {
+                        cert = new X509Certificate2(Convert.FromBase64String(signatureCertData.Current));
+                        signingKeyId = cert.Thumbprint;
+                    }
+                    catch (CryptographicException)
+                    {
+                        return new ConfigurationValidationResult
+                        {
+                            ErrorMessage = LogMessages.IDX22711,
+                            Succeeded = false
+                        };
+                    }
+                    finally
+                    {
+                        if (cert != null)
+                        {
+                            ((IDisposable)cert).Dispose();
+                        }
+                    }
+                }
+            }
+
+            // We know the key used to sign the doc is part of the token signing keys as per the spec.
+            // http://docs.oasis-open.org/wsfed/federation/v1.2/os/ws-federation-1.2-spec-os.html#_Toc223174958:~:text=%3C/fed%3ATargetScopes%20%3E-,3.1.15%20%5BSignature%5D%20Property,-The%20OPTIONAL%20%5BSignature
+            // If the metadata is for a token issuer then the key used to sign issued tokens SHOULD
+            // be used to sign this document.  This means that if a <fed:TokenSigningKey> is specified,
+            // it SHOULD be used to sign this document.
+
+            foreach (SecurityKey key in configuration.SigningKeys)
+            {
+                if (signingKeyId != key.KeyId)
+                    continue;
+
+                try
+                {
+                    configuration.Signature.Verify(key, key.CryptoProviderFactory);
```

Comment:
Looks like this can throw `ArgumentNullException` too. That would be a bit odd since in other cases where there's unexpected input it's returned as a failed `ConfigurationValidationResult`. Is this intentional?

Reply:
This is a good catch. 

On the our scenario this is not a real problem. If we deserialize the configuration using the metadataSerializer class the security key always use the default crypto provider and this value is never null.

Having said that, it is possible to modify this property before calling the validator. We need to handle that scenario.

I will update the logic to check if there is a crypto provider before attempting to verify.